File: iso27001-fastapi/app/core/rate_limiter.py
"""
A.9 / A.17: Redis-backed sliding-window rate limiter with three tiers.

  auth   — 10  req/min per IP  (login — A.9: brute-force protection)
  write  — 30  req/min per IP  (POST/PUT/PATCH/DELETE — A.17: write protection)
  global — 100 req/min per IP  (everything else — A.17: DoS protection)

Lua script ensures atomic ZREMRANGEBYSCORE + ZCARD + ZADD (no TOCTOU race).
Falls back to an in-process sliding window when Redis is unavailable.
"""

import time
from typing import Any
from fastapi import Request, HTTPException, status

_LIMITS: dict[str, int] = {
    "auth":   10,
    "write":  30,
    "global": 100,
}

_LUA_SCRIPT = """
local key = KEYS[1]
local limit = tonumber(ARGV[1])
local now = tonumber(ARGV[2])
local window = tonumber(ARGV[3])
local clear_before = now - window

redis.call('ZREMRANGEBYSCORE', key, 0, clear_before)
local count = redis.call('ZCARD', key)

if count < limit then
    redis.call('ZADD', key, now, tostring(now))
    redis.call('EXPIRE', key, window)
    return 0
else
    return 1
end
"""

# ── in-process fallback (dev / no-Redis) ─────────────────────────────────────
_local_windows: dict[str, list[float]] = {}


def _local_check(key: str, limit: int, window: int) -> bool:
    now = time.time()
    entries = [t for t in _local_windows.get(key, []) if t > now - window]
    if len(entries) >= limit:
        _local_windows[key] = entries
        return False
    entries.append(now)
    _local_windows[key] = entries
    return True


def _redis_client() -> Any:
    try:
        import redis as _redis
        from app.config.settings import settings

        client = _redis.Redis.from_url(
            settings.REDIS_URL, encoding="utf-8", decode_responses=True, socket_connect_timeout=0.5
        )
        client.ping()
        return client
    except Exception:
        return None


def _tier(request: Request) -> str:
    path = request.url.path
    method = request.method.upper()
    if "/auth/" in path or path.endswith("/auth"):
        return "auth"
    if method in {"POST", "PUT", "PATCH", "DELETE"}:
        return "write"
    return "global"


class RedisRateLimiter:
    """
    A.17: Tiered sliding-window rate limiter.
    Redis-backed with silent in-process fallback.
    """

    async def check(self, request: Request) -> None:
        tier = _tier(request)
        limit = _LIMITS[tier]
        window = 60
        client_ip = request.client.host if request.client else "unknown"
        key = f"rate_limit:{tier}:{client_ip}"
        now = time.time()

        r = _redis_client()
        if r is not None:
            allowed = int(r.eval(_LUA_SCRIPT, 1, key, limit, now, window)) == 0
        else:
            allowed = _local_check(key, limit, window)

        if not allowed:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail={"code": "RATE_LIMIT", "message": f"Rate limit exceeded ({tier}: {limit}/min)"},
            )

File: iso27001-gin/internal/core/middleware/rate_limiter.go
package middleware

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	goredis "github.com/redis/go-redis/v9"
)

// A.9 / A.17: Tiered sliding-window rate limiter.
//
//	auth   — 10  req/min per IP  (brute-force protection)
//	write  — 30  req/min per IP  (POST/PUT/PATCH/DELETE)
//	global — 100 req/min per IP  (everything else)
//
// Redis-backed with goroutine-safe in-process fallback.
var tierLimits = map[string]int{
	"auth":   10,
	"write":  30,
	"global": 100,
}

// luaScript atomically removes expired entries, counts current window, and
// conditionally adds the new entry — prevents TOCTOU race on Redis.
const luaScript = `
local key = KEYS[1]
local limit = tonumber(ARGV[1])
local now   = tonumber(ARGV[2])
local window = tonumber(ARGV[3])
local clearBefore = now - window
redis.call('ZREMRANGEBYSCORE', key, 0, clearBefore)
local count = redis.call('ZCARD', key)
if count < limit then
    redis.call('ZADD', key, now, tostring(now))
    redis.call('EXPIRE', key, window)
    return 0
else
    return 1
end`

// localWindow holds the in-process fallback sliding windows.
var (
	localMu      sync.Mutex
	localWindows = map[string][]int64{} // key → slice of unix-nano timestamps
)

func localCheck(key string, limit int, windowSec int64) bool {
	now := time.Now().UnixNano()
	cutoff := now - windowSec*int64(time.Second)
	localMu.Lock()
	defer localMu.Unlock()
	entries := localWindows[key]
	filtered := entries[:0]
	for _, t := range entries {
		if t > cutoff {
			filtered = append(filtered, t)
		}
	}
	if len(filtered) >= limit {
		localWindows[key] = filtered
		return false
	}
	localWindows[key] = append(filtered, now)
	return true
}

func tier(method, path string) string {
	if strings.Contains(path, "/auth/") || strings.HasSuffix(path, "/auth") {
		return "auth"
	}
	m := strings.ToUpper(method)
	if m == "POST" || m == "PUT" || m == "PATCH" || m == "DELETE" {
		return "write"
	}
	return "global"
}

// RateLimit returns a Gin middleware enforcing per-IP rate limits.
// Pass a non-nil redis client to use Redis; nil falls back to in-process.
func RateLimit(rdb *goredis.Client) gin.HandlerFunc {
	script := goredis.NewScript(luaScript)
	return func(c *gin.Context) {
		t := tier(c.Request.Method, c.Request.URL.Path)
		limit := tierLimits[t]
		ip := c.ClientIP()
		key := fmt.Sprintf("rate_limit:%s:%s", t, ip)
		now := time.Now().UnixNano()

		allowed := true
		if rdb != nil {
			ctx, cancel := context.WithTimeout(c.Request.Context(), 300*time.Millisecond)
			defer cancel()
			res, err := script.Run(ctx, rdb, []string{key}, limit, now, int64(60)).Int()
			if err == nil {
				allowed = res == 0
			}
			// On Redis error fall through to in-process
		}
		if rdb == nil {
			allowed = localCheck(key, limit, 60)
		}

		if !allowed {
			c.Header("X-RateLimit-Limit", fmt.Sprintf("%d", limit))
			c.Header("X-RateLimit-Remaining", "0")
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error": gin.H{
					"code":    "RATE_LIMIT",
					"message": fmt.Sprintf("Rate limit exceeded (%s: %d/min)", t, limit),
				},
			})
			return
		}
		c.Header("X-RateLimit-Limit", fmt.Sprintf("%d", limit))
		c.Next()
	}
}
